client/src/actions/annotation.ts

```diff
@@ -0,0 +1,643 @@
+
+  // If caches are identical, then nothing changed beyond new/removed columns
+  if (currentObsCache !== lastObsCache) {
+    for (const col of currentWritable) {
+      const currentHas = currentObsCache?.hasCol?.(col) ?? false;
+      const lastHas = lastObsCache?.hasCol?.(col) ?? false;
+      if (currentHas && lastHas) {
+        if (currentObsCache.col(col) !== lastObsCache.col(col))
+          changed.add(col);
+      } else if (currentHas && !lastHas) {
+        changed.add(col);
+      }
+    }
+  }
+
+  return Array.from(changed);
+}
+
+export function deletedWritableAnnotations(
+  annoMatrix: AnnoMatrix | undefined | null,
+  lastSavedAnnoMatrix: AnnoMatrix | null
+): string[] {
+  if (!annoMatrix || !lastSavedAnnoMatrix) return [];
+
+  const currentBase = annoMatrix.base();
+  const lastBase = lastSavedAnnoMatrix.base();
+
+  if (currentBase === lastBase) return [];
+
+  const currentWritable = writableAnnotations(currentBase);
+  const lastWritable = writableAnnotations(lastBase);
+
+  // Find columns that existed in last save but no longer exist
+  const deleted: string[] = [];
+  for (const col of lastWritable) {
+    if (!currentWritable.includes(col)) {
+      deleted.push(col);
+    }
+  }
+
+  return deleted;
+}
+
+export const needToSaveObsAnnotations = (
+  annoMatrix: AnnoMatrix | undefined,
+  lastSavedAnnoMatrix: AnnoMatrix | null
+): boolean =>
+  changedWritableAnnotations(annoMatrix, lastSavedAnnoMatrix).length > 0 ||
+  deletedWritableAnnotations(annoMatrix, lastSavedAnnoMatrix).length > 0;
+
+export const saveObsAnnotationsAction =
+  (): ThunkResult => async (dispatch: AppDispatch, getState: GetState) => {
+    const state = getState();
+    const { autosave } = state;
+    const { lastSavedAnnoMatrix, obsAnnotationSaveInProgress } = autosave;
+    const annoMatrix = state.annoMatrix?.base();
+
+    if (!annoMatrix) return;
+
+    if (obsAnnotationSaveInProgress || annoMatrix === lastSavedAnnoMatrix) {
+      return;
+    }
+
+    const changedCols = changedWritableAnnotations(
+      annoMatrix,
+      lastSavedAnnoMatrix
+    );
+    const deletedCols = deletedWritableAnnotations(
+      annoMatrix,
+      lastSavedAnnoMatrix
+    );
+
+    // If no changes or deletions, mark as complete
+    if (changedCols.length === 0 && deletedCols.length === 0) {
+      dispatch({
+        type: "writable obs annotations - save complete",
+        lastSavedAnnoMatrix: annoMatrix,
+      });
+      return;
+    }
+
+    // If only deletions (no changed data to save), mark as complete
+    // Backend category deletion will be handled separately via dedicated endpoints
+    if (changedCols.length === 0 && deletedCols.length > 0) {
+      dispatch({
+        type: "writable obs annotations - save complete",
+        lastSavedAnnoMatrix: annoMatrix,
+      });
+      return;
+    }
+
+    dispatch({
+      type: "writable obs annotations - save started",
+    });
+
+    const df = await annoMatrix.fetch(Field.obs, changedCols);
```

Comment:
Autosave path: changed columns -> MatrixFBS -> deflate -> octet-stream PUT to /annotations/obs with credentials. Marks complete or error accordingly.